File: app/src/main/java/com/gtm/archcomponents/notes/ui/dashboard/NotesDashBoardActivity.java
package com.gtm.archcomponents.notes.ui.dashboard;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.gtm.archcomponents.R;
import com.gtm.archcomponents.base.BaseActivity;
import com.gtm.archcomponents.notes.room.Notes;
import com.gtm.archcomponents.notes.ui.adapter.NotesAdapter;
import com.gtm.archcomponents.notes.ui.new_note.AddNewNoteActivity;

import java.util.List;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class NotesDashBoardActivity extends BaseActivity {

    private NotesAdapter notesAdapter;
    private CompositeDisposable compositeDisposable;
    private boolean isDeleted;
    private int selectedPosition;
    private NotesDashBoardViewModel notesDashBoardViewModel;
    private Menu mMenu;
    private List<Notes> mNotes;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TextView mTvHint = findViewById(R.id.id_tv_hint_no_notes);
        getSupportActionBar().setTitle(getTitle());
        setTitle(getString(R.string.app_name));
        //Create a new NOTE
        FloatingActionButton fabNewNote = findViewById(R.id.fab_new_note);
        fabNewNote.setOnClickListener(v -> startActivity(new Intent(NotesDashBoardActivity.this, AddNewNoteActivity.class)));


        // Display existing NOTES
        RecyclerView mRcvNotes = findViewById(R.id.recyclerView);
        notesDashBoardViewModel = new ViewModelProvider(this).get(NotesDashBoardViewModel.class);
        notesAdapter = new NotesAdapter(this);
        mRcvNotes.setLayoutManager(new GridLayoutManager(this, 2));
        mRcvNotes.setAdapter(notesAdapter);

        Disposable disposable = notesAdapter.getLongClickedItemPosition().subscribe(position ->
                {
                    this.selectedPosition = position;
                    showMenu();
                    Log.d("NotesDashBoardActivity", "" + position);
                }

        );

        compositeDisposable = new CompositeDisposable();
        compositeDisposable.add(disposable);
        notesDashBoardViewModel.mAllNotes.observe(this, notes -> {

            this.mNotes = notes;
            if ((notes.size() == 0)) {
                mTvHint.setVisibility(View.VISIBLE);
            } else {
                mTvHint.setVisibility(View.GONE);
            }

            notesAdapter.setNotes(notes);
        });

    }


    @Override
    public int getLayoutResource() {
        return R.layout.activity_notes_dash_board;
    }

    /**
     * Show the delete option menu
     */
    private void showMenu() {
        mMenu.findItem(R.id.menu_delete).setVisible(true);

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == R.id.menu_delete) {
            deleteSelectedNote();
        }


        return false;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_notes_list, menu);

        mMenu = menu;
        return true;
    }


    /**
     * Retrieve the data from DB and delete the selected note.
     */
    private void deleteSelectedNote() {

        Notes note = this.mNotes.get(selectedPosition);

        int noteId = note.getId();


        notesDashBoardViewModel.deleteSelectedNote(note);


    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        compositeDisposable.clear();
        compositeDisposable.dispose();

    }
}

File: app/src/main/java/com/gtm/pustak/kotlin/ui/dashboard/NotesDashBoardActivity.kt
package com.gtm.pustak.kotlin.ui.dashboard

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.gtm.pustak.R
import com.gtm.pustak.databinding.ActivityNotesDashBoardBinding
import com.gtm.pustak.kotlin.base.BaseActivity
import com.gtm.pustak.kotlin.ui.adapter.NotesAdapter
import com.gtm.pustak.kotlin.ui.new_note.AddNewNoteActivity
import com.gtm.pustak.notes.room.Notes
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.activity_notes_dash_board.*

class NotesDashBoardActivity : BaseActivity<ActivityNotesDashBoardBinding>() {

    var selectedPosition = 0
    var mNotes: List<Notes>? = null
    lateinit var mMenu: Menu
    lateinit var notesDashBoardViewModel: NotesDashBoardViewModel
    override fun getLayoutResource(): ActivityNotesDashBoardBinding {
        return ActivityNotesDashBoardBinding.inflate(layoutInflater)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        supportActionBar?.setTitle(title)
        title = getString(R.string.app_name)

        binding!!.fabNewNote.setOnClickListener {
            startActivity(Intent(this@NotesDashBoardActivity, AddNewNoteActivity::class.java))
        }


        notesDashBoardViewModel = ViewModelProvider(this).get(NotesDashBoardViewModel::class.java)

        val notesAdapter = NotesAdapter(this)
        binding!!.recyclerView.layoutManager = GridLayoutManager(this, 2)
        binding!!.recyclerView.adapter = notesAdapter


        val disposable: Disposable = notesAdapter.getlongClickedItemPosition().subscribe { position ->
            this.selectedPosition = position
            showMenu()

        }
        compositeDisposable.add(disposable)

        notesDashBoardViewModel.mAllNotes.observe(this@NotesDashBoardActivity, Observer { notes ->
            mNotes = notes



            if (notes.isEmpty()) {
                binding!!.tvHintNoNotes.visibility = View.VISIBLE
            } else {
                binding!!.tvHintNoNotes.visibility = View.GONE
            }
            notesAdapter.setNotes(notes)
        })


    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_notes_list, menu)
        if (menu != null) {
            mMenu = menu
        }
        return true
    }


    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == R.id.menu_delete) {
            deleteSelectedNote()
            hideMenu()
        }
        return true
    }

    /**
     * hide the delete menu
     */
    private fun hideMenu() {
        mMenu.findItem(R.id.menu_delete).setVisible(false)
    }

    /**
     * Show the delete option menu
     */
    private fun showMenu() {
        mMenu.findItem(R.id.menu_delete).setVisible(true)

    }


    /**
     * Retrieve the data from DB and delete the selected note.
     */
    fun deleteSelectedNote() {
        var notes = mNotes!!.get(selectedPosition)
        notesDashBoardViewModel.deleteSelectedNote(notes)
    }

}
